### ISSUE TITLE
Centralize knowledge of the specific ConfigStorageInterface

### ISSUE THREAD

**Issue by sentaidigital** [backdrop/backdrop-issues#2275]
A number of functions call `new ConfigStorageInterface($directory)` with either the active or staging directories.  These should be reduced down to one function that creates the ConfigFileInterface object.

**Comment by sentaidigital** [comment]
The existing unit tests cover this code, including the new `ConfigStorageInterface::importArchive()` (handled by `ConfigurationUITest::testImport()`).  I don't see any unit tests specifically testing `ConfigStorageInterface`, otherwise I'd have added a test there.

**Comment by quicksketch** [comment]
The PR looks really good. Additional tests to cover the missing items would be greatly appreciated.

I know I had suggested to rename `config_get_config_storage()` to `config_get_storage()`, but the current name keeps it inline with `config_get_config_directory()`. Do you have a preference for one name over the other?

Let's tag this for 1.6.0.

**Comment by sentaidigital** [comment]
Keeping it inline with `config_get_config_directory()` was the reason for `config_get_config_storage()`, as redundant as it sounds.  I have a mild preference for `config_get_config_storage()`.

**Comment by sentaidigital** [comment]
The latest version of this patch incorporates the `ConfigStorageInterface` API additions from #2277 and applies these new APIs in the core code.

**Comment by sentaidigital** [comment]
Unit tests for testing the Config object have been added to the patch.

**Comment by jenlampton** [comment]
I removed the 1.6.0 tag because I think we need to consider the implications of adding an edge-case feature more carefully. We haven't even had a single release since this idea was introduced, and I'd like to give it some more time to see if we can get some more feedback.

**Comment by sentaidigital** [comment]
While it does add some API, this PR doesn't add any new functionality.  Instead cleans up the `Config` object class to remove from the rest of Backdrop the assumption that the concrete `ConfigStorageInterface` class will be `ConfigFileStorage`.  For example, it adds the member function `ConfigStorageInterface::isInitialized()`, so we don't call `is_dir(config_get_config_directory('active'))` in update.inc.

This PR makes the PR for #2277 a smaller and cleaner PR focused on the `ConfigDatabaseStorage` implementation.

**Comment by jenlampton** [comment]
Tagging again for 1.6 and marking as a task (indicating that we can get this in post Jan 1st, if ready).  You are correct that this is not a feature addition and if it makes maintaining patches like #2277 easier, and @quicksketch approves, then we should get it in when it's ready :)

**Comment by quicksketch** [comment]
This PR looks fantastic. Although it mostly just moves code around, it creates a more consistent way of interacting with config storage. I did not like our hard-coded use of `file_scan_directory()` and `is_file()` for configuration checks, especially when all other actions around reading configuration was already separated out.

PR is passing all tests and looks great. It's sets us up for alternative storages if we decide to pursue that route.

I've merged https://github.com/backdrop/backdrop/pull/1612 into 1.x for 1.6.0. Thanks @jlfranklin!
